### PR TITLE
FMWK-333 Refactor repository queries

### PR DESCRIPTION
**Pull request by agrgr**
- add dedicated query creators for id, simple property, Collection, Map and POJO
- simplify repository queries API to make it more intuitive and easier to understand and validate:
-- limit repository queries to work with one parameter (no automatic AND concatenating)
-- limit repository queries to one level (not working with elements of Array, Collection or Map) except CONTAINING queries
-- add KEY_VALUE_PAIR query criterion, remove VALUE_CONTAINING
-- reorder arguments in "find by Map containing" repository query
- refactor validations
- refactor location of Qualifier classes, add Qualifier field enum
- update tests
- renaming, cleanup etc.

### CODE REVIEW

File: src/main/java/org/springframework/data/aerospike/query/FilterOperation.java [hunk 1]
```diff
@@ -946,43 +927,44 @@ public Exp filterExp(Map<String, Object> qualifierMap) {
...
-                case MAP -> Exp.val((Map<?, ?>) getValue1(qualifierMap).getObject());
+    COLLECTION_VAL_CONTAINING {
+        @Override
+        public Exp filterExp(Map<QualifierField, Object> qualifierMap) {
+            Value val = getValue(qualifierMap);
+//            // boolean values are read as BoolIntValue (INTEGER ParticleType) if Value.UseBoolBin == false
+//            // converting to BooleanValue to process correctly
+//            if (val instanceof Value.BoolIntValue) {
+//                qualifierMap.put(VALUE, new Value.BooleanValue((Boolean) (getValue(qualifierMap).getObject())));
+//            }
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/query/FilterOperation.java:956]: Clean the commented out.

File: src/main/java/org/springframework/data/aerospike/query/FilterOperation.java [hunk 2]
```diff
@@ -991,33 +973,34 @@ public Exp filterExp(Map<String, Object> qualifierMap) {
...
-                case MAP -> Exp.val((Map<?, ?>) getValue1(qualifierMap).getObject());
+    COLLECTION_VAL_NOT_CONTAINING {
+        @Override
+        public Exp filterExp(Map<QualifierField, Object> qualifierMap) {
+            Value val = getValue(qualifierMap);
+//            // boolean values are read as BoolIntValue (INTEGER ParticleType) if Value.UseBoolBin == false
+//            // so converting to BooleanValue to process correctly
+//            if (val instanceof Value.BoolIntValue) {
+//                qualifierMap.put(VALUE, new Value.BooleanValue((Boolean) (getKey(qualifierMap).getObject())));
+//            }
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/query/FilterOperation.java:992]: Clean out.

File: src/main/java/org/springframework/data/aerospike/query/qualifier/IQualifierBuilder.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package org.springframework.data.aerospike.query.qualifier;
+
+import java.util.Map;
+
+public interface IQualifierBuilder {
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/query/qualifier/IQualifierBuilder.java:5]: Add a Javadoc with a brief description.

File: src/main/java/org/springframework/data/aerospike/repository/query/IAerospikeQueryCreator.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package org.springframework.data.aerospike.repository.query;
+
+import org.springframework.data.aerospike.query.qualifier.Qualifier;
+
+public interface IAerospikeQueryCreator {
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/repository/query/IAerospikeQueryCreator.java:5]: Add a Javadoc.

File: src/main/java/org/springframework/data/aerospike/repository/query/QueryParameters.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+    public QueryParameters(Iterator<?> parameters) {
+//        List<Object> params = new ArrayList<>();
+//
+//        Object value1 = null;
+//        if (parameters.hasNext()) {
+//            value1 = parameters.next();
+//        }
+//        params.add(convertIfNecessary(value1));
+//        parameters.forEachRemaining(params::add);
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/repository/query/QueryParameters.java:24]: Should be empty?

File: src/test/java/org/springframework/data/aerospike/repository/query/findBy/noindex/EqualsTests.java [hunk 6]
```diff
@@ -135,6 +138,119 @@ void findByNestedSimplePropertyEqualsNegativeTest() {
             .hasMessage("Address.zipCode EQ: invalid number of arguments, expecting one");
     }
 
+    @Test
+        // find by deeply nested String POJO field
```
Comment (reugn) [inline src/test/java/org/springframework/data/aerospike/repository/query/findBy/noindex/EqualsTests.java:142]: Indentation

File: src/test/java/org/springframework/data/aerospike/repository/query/findBy/noindex/EqualsTests.java [hunk 7]
```diff
@@ -135,6 +138,119 @@ void findByNestedSimplePropertyEqualsNegativeTest() {
...
+        List<Person> result =
+            repository.findByFriendFriendFriendFriendFriendFriendFriendFriendBestFriendFriendAddressZipCode(zipCode);
+
+        assertThat(result).containsExactly(douglas);
+
+        TestUtils.setFriendsToNull(repository, allPersons.toArray(Person[]::new)); // cleanup
+    }
+
+    @Test
+        // find by deeply nested Integer POJO field
```
Comment (reugn) [inline src/test/java/org/springframework/data/aerospike/repository/query/findBy/noindex/EqualsTests.java:179]: Indentation

File: src/test/java/org/springframework/data/aerospike/repository/query/findBy/noindex/EqualsTests.java [hunk 8]
```diff
@@ -135,6 +138,119 @@ void findByNestedSimplePropertyEqualsNegativeTest() {
...
+        assertThat(result).containsExactly(matias);
+
+        // cleanup
+        TestUtils.setFriendsToNull(repository, allPersons.toArray(Person[]::new));
+        alicia.setAddress(null);
+        repository.save(alicia);
+    }
+
+    @Test
+        // find by deeply nested POJO
```
Comment (reugn) [inline src/test/java/org/springframework/data/aerospike/repository/query/findBy/noindex/EqualsTests.java:217]: Indentation

File: src/main/java/org/springframework/data/aerospike/repository/query/MapQueryCreator.java [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+                    setQualifierBuilderValue(qb, queryParameters.get(1));
+                }
+                case VALUE -> {
+                    op = valuesOp;
+                    setQualifierBuilderValue(qb, queryParameters.get(1));
+                }
+                default -> throw new UnsupportedOperationException("Unsupported parameter: " + queryCriterion);
+            }
+        }
+        return setQualifier(converter, qb, fieldName, op, part, dotPath);
```
Comment (roimenashe) [inline src/main/java/org/springframework/data/aerospike/repository/query/MapQueryCreator.java:228]: dotPath is always null in this case, no need to create a variable

File: src/main/java/org/springframework/data/aerospike/repository/query/MapQueryCreator.java [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+            setQualifierBuilderValue(qb, queryParameters.get(0));
+            if (queryParameters.size() >= 2) setQualifierBuilderSecondValue(qb, queryParameters.get(1));
+            qualifier = setQualifier(converter, qb, fieldName, filterOperation, part, null);
+            return qualifier;
+        }
+
+        if (paramsSize == 2) {
+            qualifier = processMapTwoParams(part, queryParameters, filterOperation, fieldName);
+        } else if (queryParameters.size() < 2) {
+            if (filterOperation != FilterOperation.BETWEEN) {
```
Comment (roimenashe) [inline src/main/java/org/springframework/data/aerospike/repository/query/MapQueryCreator.java:183]: Always true (warnings from IntelliJ are not always correct but seems like that's the case here)

File: src/main/java/org/springframework/data/aerospike/repository/query/PojoQueryCreator.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+        if (isNested) { // POJO field
+            // getting MAP_VAL_ operation because the property is in a POJO which is represented by a Map in DB
+            op = getCorrespondingMapValueFilterOperationOrFail(filterOperation);
+            setQualifierBuilderKey(qb, property.getFieldName());
+            setQualifierBuilderValue(qb, queryParameters.get(0));
+            dotPath = List.of(part.getProperty().toDotPath());
+        } else { // first level POJO
+            if (op != FilterOperation.BETWEEN) {
+                if (queryParameters.size() >= 1) setQualifierBuilderValue(qb, queryParameters.get(0));
```
Comment (roimenashe) [inline src/main/java/org/springframework/data/aerospike/repository/query/PojoQueryCreator.java:114]: `!queryParameters.isEmpty()` is simpler

File: src/main/java/org/springframework/data/aerospike/repository/query/SimplePropertyQueryCreator.java [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    private void validateSimplePropertyQueryBetween(String queryPartDescription, List<Object> queryParameters) {
+        // Number of arguments is not two
+        if (queryParameters.size() != 2) {
+            throw new IllegalArgumentException(queryPartDescription + ": invalid number of arguments, expecting two");
+        }
+    }
+
+    private void validateSimplePropertyQueryIsNull(String queryPartDescription, List<Object> queryParameters) {
+        // Number of arguments is not zero
+        if (queryParameters.size() != 0) {
```
Comment (roimenashe) [inline src/main/java/org/springframework/data/aerospike/repository/query/SimplePropertyQueryCreator.java:105]: `!queryParameters.isEmpty()` is simpler

File: src/main/java/org/springframework/data/aerospike/repository/query/SimplePropertyQueryCreator.java [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+        FilterOperation op = filterOperation;
+        if (isNested) { //  POJO field
+            if (filterOperation == IS_NOT_NULL || filterOperation == IS_NULL) {
+                setQualifierBuilderValue(qb, property.getFieldName());
+            }
+
+            // getting MAP_VAL_ operation because the property is in a POJO which is represented by a Map in DB
+            op = getCorrespondingMapValueFilterOperationOrFail(op);
+
+            if (queryParameters.size() >= 1) setQualifierBuilderValue(qb, queryParameters.get(0));
```
Comment (roimenashe) [inline src/main/java/org/springframework/data/aerospike/repository/query/SimplePropertyQueryCreator.java:134]: `!queryParameters.isEmpty()`

File: src/main/java/org/springframework/data/aerospike/utility/Utils.java [hunk 14]
```diff
@@ -151,4 +152,8 @@ public static boolean isSimpleValueType(Class<?> type) {
     public static boolean isBoolean(Class<?> type) {
         return Boolean.TYPE.equals(type) || Boolean.class.equals(type);
     }
+
+    public static boolean hasNoElementsOfClass(Class<?> clazz, Collection<Object> collection) {
```
Comment (roimenashe) [inline src/main/java/org/springframework/data/aerospike/utility/Utils.java:156]: Seems like this class contains a few unused method including this new one - is it an IDE error? do we actually use them? if not I guess we can cleanup

**Review by roimenashe — APPROVED** [review]

**Review by reugn — APPROVED** [review]

**Pull request merged**